### PR TITLE
#10 create RoleResource services

### PR DESCRIPTION
**Pull request by MaryR-code**
Create RoleResource services

Get list of Roles
Create a new Role

Still no idea how to set access only for admin...

### CODE REVIEW

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Inject
+    UserControl userControl;
+
+    @Inject
+    @Claim("uid")
+    String userIdClaim;
+
+    @POST
+    //@RolesAllowed({"admin"})
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/boundary/RoleResource.java:33]: Why is it commented out? Looks correct

File: src/main/java/guru/bug/courses/srs/control/RoleControl.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    @PersistenceContext
+    EntityManager em;
+
+    @Inject
+    RoleDAO roleDAO;
+
+    public RoleEntity createRole(RoleEntity role) {
+        var id = UUID.randomUUID();
+        role.setId(id);
+        em.persist(role);
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/control/RoleControl.java:29]: There is a RoleDAO with the same code. It is better to call it.

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    @Consumes(MediaType.APPLICATION_JSON)
+    public RoleEntity setRole(@Valid @NotNull RoleEntity role) {
+        LOG.info("Creating new role...");
+        return roleControl.createRole(role);
+    }
+
+    @GET
+    //@RolesAllowed({"admin"})
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/boundary/RoleResource.java:41]: that should be right (if uncomment)

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Inject
+    RoleControl roleControl;
+
+    @Inject
+    UserControl userControl;
+
+    @Inject
+    @Claim("uid")
+    String userIdClaim;
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/boundary/RoleResource.java:30]: variable isn't in use - can be deleted

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+@Path("/roles")
+public class RoleResource {
+    private static final Logger LOG = LoggerFactory.getLogger(RoleResource.class);
+
+    @Inject
+    RoleControl roleControl;
+
+    @Inject
+    UserControl userControl;
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/boundary/RoleResource.java:26]: variable isn't in use - can be deleted

**Review by dima-fedotovs — CHANGES_REQUESTED** [review]
cleanup is required (see comments)

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java (hunk 1 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/boundary/RoleResource.java]: I used finaly @PermitAll

File: src/main/java/guru/bug/courses/srs/control/RoleControl.java (hunk 2 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/control/RoleControl.java]: done

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java (hunk 4 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/boundary/RoleResource.java]: done

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java (hunk 3 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/boundary/RoleResource.java]: +

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java (hunk 5 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/boundary/RoleResource.java]: +

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        }
+        return roleObj;
+    }
+
+    @GET
+    @PermitAll
+    @Consumes(MediaType.APPLICATION_JSON)
+    public List<RoleEntity> getRoles(@Context SecurityContext sec) {
+        List<RoleEntity> roleList = new ArrayList<>();
+        if (sec.isUserInRole("admin")) {
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/boundary/RoleResource.java:45]: @RolesAllowed is more preferable.

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            roleObj = roleControl.createRole(role);
+            LOG.info("Creating new role...");
+        }
+        return roleObj;
+    }
+
+    @GET
+    @PermitAll
+    @Consumes(MediaType.APPLICATION_JSON)
+    public List<RoleEntity> getRoles(@Context SecurityContext sec) {
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/boundary/RoleResource.java:43]: please add @Produces annotation.

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    @POST
+    @PermitAll
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public RoleEntity setRole(@Valid @NotNull RoleEntity role,
+                              @Context SecurityContext sec) {
+        var roleObj = new RoleEntity();
+        if (sec.isUserInRole("admin")) {
+            roleObj = roleControl.createRole(role);
+            LOG.info("Creating new role...");
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/boundary/RoleResource.java:35]: It is ok to log the creation of the role to the INFO level. Would be more useful adding a name of the created role to the log as well.

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @GET
+    @PermitAll
+    @Consumes(MediaType.APPLICATION_JSON)
+    public List<RoleEntity> getRoles(@Context SecurityContext sec) {
+        List<RoleEntity> roleList = new ArrayList<>();
+        if (sec.isUserInRole("admin")) {
+            roleList = roleControl.findAll();
+            LOG.info("Selecting list of roles...");
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/boundary/RoleResource.java:47]: I suggest using DEBUG level in this case. I mean nobody wants to know how often and how many times roles were selected in production. But could be useful when debugging.

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    private static final Logger LOG = LoggerFactory.getLogger(RoleResource.class);
+
+    @Inject
+    RoleControl roleControl;
+
+    @POST
+    @PermitAll
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public RoleEntity setRole(@Valid @NotNull RoleEntity role,
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/boundary/RoleResource.java:30]: the name of the method doesn't correspond to real action. Should be "addRole" or "createRole"

**Review by dima-fedotovs — CHANGES_REQUESTED** [review]
Some small fixes should be done before merging. See comments.

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java (hunk 10 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/boundary/RoleResource.java]: changed "createRole"

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+                              @Context SecurityContext sec) {
+        var roleObj = new RoleEntity();
+        if (sec.isUserInRole("admin")) {
```
[the remark this answers is not in the document]
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/boundary/RoleResource.java]: done

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java (hunk 8 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/boundary/RoleResource.java]: added a name of the created role

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java (hunk 7 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/boundary/RoleResource.java]: done

File: src/main/java/guru/bug/courses/srs/boundary/RoleResource.java (hunk 6 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/boundary/RoleResource.java]: changed

**Review by dima-fedotovs — APPROVED** [review]

**Comment by dima-fedotovs** [comment]
Seems everything is alright. Additional alignment could be done in separate tasks (will add).

**Pull request merged**